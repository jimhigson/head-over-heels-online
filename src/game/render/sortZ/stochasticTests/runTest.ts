import type { DrawOrderComparable } from "../DrawOrderComparable";
import type { ZGraph } from "../GraphEdges";

import { GridSpatialIndex } from "../../../physics/gridSpace/GridSpatialIndex";
import { updateZEdges } from "../updateZEdges";

export const itemCount = 100;
export const movePercentage = 0.2;
export const frameCount = 100; // round number means divide answer by 100 to get time per frame

type TestItem = DrawOrderComparable & { id: string };
/**
 * generates test items in a grid pattern for benchmarking
 */
const generateItems = (count: number): Record<string, TestItem> => {
  const items: Record<string, TestItem> = {};
  const gridSize = Math.ceil(Math.sqrt(count));

  for (let i = 0; i < count; i++) {
    const x = (i % gridSize) * 2;
    const y = Math.floor(i / gridSize) * 2;
    // deterministic z based on position - creates a wave pattern - values 0 to 5 inclusive
    // (all integers)
    const z = Math.round(Math.sin(i * 0.5) * 2.5 + 2.5);

    const id = `item-${i}`;
    items[id] = {
      id,
      state: {
        position: { x, y, z },
      },
      aabb: { x: 1, y: 1, z: 1 },
      renderAabb: { x: 1, y: 1, z: 1 },
      renderAabbOffset: { x: 0, y: 0, z: 0 },
      fixedZIndex: undefined,
    };
  }

  return items;
};
export const runTest = () => {
  // generate items
  const items = generateItems(itemCount);

  // first call with all items to build initial graph
  const itemArray = Object.values(items);
  const spatialIndex = new GridSpatialIndex(itemArray);
  const zEdgesGraph: ZGraph<string> = updateZEdges(items, spatialIndex);

  // simulate frames:
  for (let f = 0; f < frameCount; f++) {
    // select items to move
    const movedItems = new Set<TestItem>();

    // deterministically select items to move based on movePercentage
    const stepSize = Math.round(1 / movePercentage);
    for (let i = 0; i < itemArray.length; i += stepSize) {
      const item = itemArray[i];
      // move the item up
      const z = Math.round(Math.sin(i + f * 0.5) * 2.5 + 2.5);
      item.state.position.z = z;
      movedItems.add(item);
    }

    // benchmark the incremental update
    updateZEdges(items, spatialIndex, movedItems, zEdgesGraph);
  }
  return zEdgesGraph;
};
