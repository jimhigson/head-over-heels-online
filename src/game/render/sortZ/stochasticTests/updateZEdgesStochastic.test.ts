import { expect, test } from "vitest";

import { runTest } from "./runTest";

test("updateZEdges correctness", () => {
  // this test can be used when the implementation changes to verify that the output is the same
  expect(runTest()).toMatchSnapshot();
});
