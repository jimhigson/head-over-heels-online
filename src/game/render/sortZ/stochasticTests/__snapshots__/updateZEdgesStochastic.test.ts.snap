// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`updateZEdges correctness 1`] = `
Map {
  "item-22" => Map {
    "item-11" => false,
  },
  "item-12" => Map {
    "item-1" => false,
  },
  "item-23" => Map {
    "item-1" => false,
    "item-12" => false,
  },
  "item-13" => Map {
    "item-2" => false,
  },
  "item-24" => Map {
    "item-2" => false,
    "item-13" => false,
  },
  "item-14" => Map {
    "item-3" => false,
  },
  "item-36" => Map {
    "item-14" => false,
  },
  "item-37" => Map {
    "item-26" => false,
    "item-15" => false,
  },
  "item-48" => Map {
    "item-26" => false,
    "item-37" => false,
    "item-15" => false,
  },
  "item-38" => Map {
    "item-27" => false,
  },
  "item-49" => Map {
    "item-27" => false,
    "item-38" => false,
  },
  "item-39" => Map {
    "item-28" => false,
  },
  "item-47" => Map {
    "item-36" => false,
  },
  "item-62" => Map {
    "item-51" => false,
    "item-40" => false,
  },
  "item-73" => Map {
    "item-51" => false,
    "item-62" => false,
    "item-40" => false,
  },
  "item-52" => Map {
    "item-41" => false,
  },
  "item-63" => Map {
    "item-52" => false,
  },
  "item-74" => Map {
    "item-52" => false,
    "item-63" => false,
  },
  "item-64" => Map {
    "item-53" => false,
  },
  "item-86" => Map {
    "item-53" => false,
    "item-64" => false,
  },
  "item-72" => Map {
    "item-61" => false,
  },
  "item-87" => Map {
    "item-76" => false,
    "item-65" => false,
  },
  "item-98" => Map {
    "item-76" => false,
    "item-87" => false,
    "item-65" => false,
  },
  "item-77" => Map {
    "item-66" => false,
  },
  "item-88" => Map {
    "item-77" => false,
  },
  "item-99" => Map {
    "item-77" => false,
    "item-88" => false,
  },
  "item-89" => Map {
    "item-78" => false,
  },
  "item-25" => Map {
    "item-3" => false,
    "item-14" => false,
  },
  "item-75" => Map {
    "item-53" => false,
    "item-42" => false,
  },
  "item-51" => Map {
    "item-40" => false,
  },
  "item-26" => Map {
    "item-15" => false,
  },
  "item-76" => Map {
    "item-65" => false,
  },
  "item-21" => Map {
    "item-10" => false,
  },
  "item-46" => Map {
    "item-35" => false,
  },
  "item-55" => Map {
    "item-44" => false,
  },
  "item-96" => Map {
    "item-85" => false,
  },
}
`;
